File: utils/youtube.jl
function get_recent_videos(channel_id::String; max_results::Int=10)
    # Get the uploads playlist ID
    channel_url = "https://www.googleapis.com/youtube/v3/channels"
    channel_params = Dict(
        "part" => "contentDetails",
        "id" => channel_id,
        "key" => GOOGLE_API_KEY
    )
    
    channel_response = HTTP.get(channel_url; query=channel_params)
    channel_data = JSON.parse(String(channel_response.body))
    
    if isempty(channel_data["items"])
        error("Channel not found or invalid channel ID")
    end
    uploads_playlist_id = channel_data["items"][1]["contentDetails"]["relatedPlaylists"]["uploads"]
    
    # Get videos from the uploads playlist
    playlist_url = "https://www.googleapis.com/youtube/v3/playlistItems"
    playlist_params = Dict(
        "part" => "snippet",
        "playlistId" => uploads_playlist_id,
        "maxResults" => string(max_results),
        "key" => GOOGLE_API_KEY
    )
    
    playlist_response = HTTP.get(playlist_url; query=playlist_params)
    playlist_data = JSON.parse(String(playlist_response.body))
    
    # Extract video information
    videos = []
    for item in playlist_data["items"]
        snippet = item["snippet"]
        video_info = Dict(
            "video_id" => snippet["resourceId"]["videoId"],
            "title" => snippet["title"],
            "description" => snippet["description"],
            "published_at" => snippet["publishedAt"],
            "url" => "https://www.youtube.com/watch?v=$(snippet["resourceId"]["videoId"])"
        )
        push!(videos, video_info)
    end
    
    return videos
end

function parse_youtube_date(date_string::String)
    # YouTube returns dates in ISO 8601 format: "2022-01-13T02:32:43Z"
    # Remove the 'Z' if present and parse
    clean_date = replace(date_string, "Z" => "")
    
    # Handle both with and without milliseconds
    if occursin(".", clean_date)
        return DateTime(clean_date, dateformat"yyyy-mm-ddTHH:MM:SS.s")
    else
        return DateTime(clean_date, dateformat"yyyy-mm-ddTHH:MM:SS")
    end
end

function day_with_suffix(day::Int)
    if day in [11, 12, 13]
        return "$(day)th"
    end
    
    last_digit = day % 10
    
    if last_digit == 1
        return "$(day)st"
    elseif last_digit == 2
        return "$(day)nd"
    elseif last_digit == 3
        return "$(day)rd"
    else
        return "$(day)th"
    end
end

function format_date_readable(date_string::String; use_suffix::Bool=false)
    dt = parse_youtube_date(date_string)
    
    # Get month abbreviation
    month_abbrev = Dates.format(dt, "u")  # Short month name (Jan, Feb, etc.)
    
    # Format day with or without suffix
    day_str = use_suffix ? day_with_suffix(Dates.day(dt)) : string(Dates.day(dt))
    
    # Format as "Mon. DD, YYYY"
    return "$(month_abbrev). $(day_str), $(Dates.year(dt))"
end

function process_videos(videos)
    video_markdown = ""
    for video in videos 
        local date = format_date_readable(video["published_at"]; use_suffix=true)
        local title = video["title"]
        local url = video["url"]
        local description = video["description"]
        video_markdown *= """\
            \\emoji{clock}  **[$date] $title:** [View on Youtube]($url)\\
            $description\\
            \\
        """
    end 
    return video_markdown
end

File: sources/youtube.ts
import { GOOGLE_API_KEY } from "./commands.ts";
const GLOBAL_URL_PATTERN = /(https?:\/\/[^\s]+)/g;
const GLOBAL_TAG_PATTERN = /#(\w+)/g;

interface VideoInfo {
  video_id: string;
  title: string;
  description: string;
  published_at: string;
  url: string;
}

export async function getRecentVideos(
  channelId: string,
  maxResults: number = 10,
): Promise<VideoInfo[]> {
  // Get the uploads playlist ID
  const channelUrl = "https://www.googleapis.com/youtube/v3/channels";
  const channelParams = new URLSearchParams({
    part: "contentDetails",
    id: channelId,
    key: GOOGLE_API_KEY!,
  });

  const channelResponse = await fetch(`${channelUrl}?${channelParams}`);
  const channelData = await channelResponse.json();

  if (!channelData.items || channelData.items.length === 0) {
    throw new Error("Channel not found or invalid channel ID");
  }

  const uploadsPlaylistId =
    channelData.items[0].contentDetails.relatedPlaylists.uploads;

  // Get videos from the uploads playlist
  const playlistUrl = "https://www.googleapis.com/youtube/v3/playlistItems";
  const playlistParams = new URLSearchParams({
    part: "snippet",
    playlistId: uploadsPlaylistId,
    maxResults: maxResults.toString(),
    key: GOOGLE_API_KEY!,
  });

  const playlistResponse = await fetch(`${playlistUrl}?${playlistParams}`);
  const playlistData = await playlistResponse.json();

  // Extract video information
  const videos: VideoInfo[] = [];
  for (const item of playlistData.items) {
    const snippet = item.snippet;
    const videoInfo: VideoInfo = {
      video_id: snippet.resourceId.videoId,
      title: snippet.title,
      description: snippet.description,
      published_at: snippet.publishedAt,
      url: `https://www.youtube.com/watch?v=${snippet.resourceId.videoId}`,
    };
    videos.push(videoInfo);
  }

  return videos;
}

function parseYoutubeDate(dateString: string): Date {
  // YouTube returns dates in ISO 8601 format: "2022-01-13T02:32:43Z"
  // Parse ISO 8601 format directly
  return new Date(dateString);
}

function dayWithSuffix(day: number): string {
  if ([11, 12, 13].includes(day)) {
    return `${day}th`;
  }

  const lastDigit = day % 10;

  if (lastDigit === 1) {
    return `${day}st`;
  } else if (lastDigit === 2) {
    return `${day}nd`;
  } else if (lastDigit === 3) {
    return `${day}rd`;
  } else {
    return `${day}th`;
  }
}

function formatDateReadable(
  dateString: string,
  useSuffix: boolean = false,
): string {
  const dt = parseYoutubeDate(dateString);

  // Get month abbreviation
  const monthAbbrev = dt.toLocaleDateString("en-US", { month: "short" });

  // Format day with or without suffix
  const dayStr = useSuffix
    ? dayWithSuffix(dt.getDate())
    : dt.getDate().toString();

  // Format as "Mon. DD, YYYY"
  return `${monthAbbrev}. ${dayStr}, ${dt.getFullYear()}`;
}

export function processVideos(videos: VideoInfo[]): string {
  let videoMarkdown = "";
  let index = 0;
  for (const video of videos) {
    const date = formatDateReadable(video.published_at, true);
    const title = video.title;
    const url = video.url;
    let description = video.description;
    description = description.replace(GLOBAL_URL_PATTERN, "[$1]($1)");
    description = description.replace(
      GLOBAL_TAG_PATTERN,
      "[#$1](https://www.youtube.com/hashtag/$1)",
    );
    videoMarkdown += `\\emoji{clock} **[${date}] ${title}:** [View on Youtube](${url})\\\n${description}`;
    if (index !== videos.length - 1) {
      videoMarkdown += `\\\n\\\n`;
    }
    index += 1;
  }
  return videoMarkdown;
}
